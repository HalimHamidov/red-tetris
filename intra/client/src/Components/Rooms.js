import React from "react";
import { useState, useEffect } from "react";
import { toast } from "react-toastify";
import axios from "axios";
import { useSelector, useDispatch } from "react-redux";
import {
  StyledButton,
  StyledInput,
  RoomWrapper,
  StyledSelect,
  InputsWrapper,
  FormWrapper,
  TABLE,
  THEAD,
  TR,
  TD,
  TBODY,
  TH,
  JoinButton,
} from "./styling/StyledForm";
import { UPDATE_MODE } from "../actions/roomAction";

const Rooms = () => {
  const dispatch = useDispatch();
  const [rooms, setRooms] = useState([]);
  const [Room, setRoomname] = useState("");
  const [mode, setMode] = useState("Solo");
  let playerState = useSelector((state) => {
    return state.player;
  });

  function addRoom(e) {
    e.preventDefault();
    if (Room.trim() !== "") {
      window.location.hash = `${Room.trim()}[${playerState.username}]`;
      dispatch({ type: UPDATE_MODE, data: mode });
    } else {
      setRoomname("");
      toast.error(
        "Please enter 12 characters in length correct Player and Room names"
      );
    }
  }

  useEffect(() => {
    axios.get(`${process.env.REACT_APP_API_URL}/getRooms`).then((res) => {
      setRooms([...res.data]);
    });
    return () => {
      setRooms([]);
    };
  }, []);

  return (
    <RoomWrapper onSubmit={(e) => addRoom(e)}>
      <FormWrapper>
        <InputsWrapper>
          <StyledInput
            placeholder="Type your Room Name"
            onChange={(e) => setRoomname(e.target.value)}
          />{" "}
          <StyledSelect value={mode} onChange={(e) => setMode(e.target.value)}>
            <option> Solo </option> <option> Multi - Player </option>{" "}
          </StyledSelect>{" "}
        </InputsWrapper>{" "}
        <StyledButton type="submit"> START </StyledButton>{" "}
      </FormWrapper>{" "}
      {rooms.length > 0 && (
        <TABLE>
          <THEAD>
            <TR>
              <TH> Room - Name </TH> <TH> Members </TH> <TH> Mode </TH>{" "}
              <TH> </TH>{" "}
            </TR>{" "}
          </THEAD>{" "}
          <TBODY>
            {" "}
            {rooms.map(
              (item, i) =>
                item.room.length > 0 && (
                  <TR key={i}>
                    <TD>
                      <strong> {item.room} </strong>{" "}
                    </TD>{" "}
                    <TD>
                      {" "}
                      {item.members}
                      /5
                    </TD>
                    <TD> {item.mode} </TD>{" "}
                    <TD>
                      <JoinButton
                        onClick={() =>
                          (window.location = `${process.env.REACT_APP_FRONTEND_URL}/#${item.room}[${playerState.username}]`)
                        }
                      >
                        Join{" "}
                      </JoinButton>{" "}
                    </TD>{" "}
                  </TR>
                )
            )}{" "}
          </TBODY>{" "}
        </TABLE>
      )}{" "}
    </RoomWrapper>
  );
};

export default Rooms;
