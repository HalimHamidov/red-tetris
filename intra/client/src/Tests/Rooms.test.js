import renderer from "react-test-renderer";
import Rooms from "../Components/Rooms";
import store from "../Store";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";
import { Provider } from "react-redux";
import * as reactRedux from "react-redux";
import { render, fireEvent, screen } from "@testing-library/react";

const mockStore = configureMockStore([thunk]);

describe("Rooms", () => {
  const useDispatchMock = jest.spyOn(reactRedux, "useDispatch");

  const res = {
    data: [
      {
        room: "testRoom",
        members: 1,
        mode: "Solo",
      },
    ],
  };

  beforeEach(() => {
    useDispatchMock.mockClear();
  });

  test("Rooms render test ", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <Rooms />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Test form submit with wrong value", () => {
    const store = mockStore({
      player: { username: "testUser" },
    });
    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <Rooms />
      </Provider>
    );
    const roomInput = getByPlaceholderText(/Type your Room Name/);

    fireEvent.change(roomInput, { target: { value: "" } });
    fireEvent.click(getByText(/START/));

    const errMsg =
      "[ERROR] Player and Room names must be 1 to 12 alphanumeric characters in length";
    expect(screen.findByText(errMsg)).toBeDefined();
  });

  test("Test form submit with correct value", () => {
    const store = mockStore({
      player: { username: "testUser" },
    });
    useDispatchMock.mockReturnValue(jest.fn());
    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <Rooms />
      </Provider>
    );
    const roomInput = getByPlaceholderText(/Type your Room Name/);

    fireEvent.change(roomInput, { target: { value: "testRoom" } });
    fireEvent.click(getByText(/START/));
    expect(useDispatchMock).toHaveBeenCalled();
    global.window = Object.create(window);
    const url = "#testRoom[testUser]";
    Object.defineProperty(window, "location", {
      value: {
        hash: url,
      },
    });
    expect(window.location.hash).toEqual(url);
  });
});
