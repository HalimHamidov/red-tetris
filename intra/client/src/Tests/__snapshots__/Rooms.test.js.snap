// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rooms Rooms render test  1`] = `
<div
  className="sc-hKwDye tCDCV"
  onSubmit={[Function]}
>
  <form
    className="sc-eCImPb bQyJNu"
  >
    <div
      className="sc-jRQBWg eklnKY"
    >
      <input
        className="sc-bdvvtL bEPCcd"
        onChange={[Function]}
        placeholder="Type your Room Name"
      />
       
      <select
        className="sc-gKclnd jXdwOV"
        onChange={[Function]}
        value="Solo"
      >
        <option>
           Solo 
        </option>
         
        <option>
           Multi - Player 
        </option>
         
      </select>
       
    </div>
     
    <button
      className="sc-gsDKAQ gUImZr"
      type="submit"
    >
       START 
    </button>
     
  </form>
   
   
</div>
`;
